e2e/playwright/support/kv.ts

```diff
@@ -0,0 +1,64 @@
+import {Client} from 'pg';
+
+export function getClient() {
+    const client = new Client({
+        host: 'localhost',
+        user: 'mmuser',
+        password: 'mostest',
+        database: 'mattermost_test',
+    });
+
+    client.connect();
+    client.end = client.end.bind(client);
+    return client;
+}
+
+export const getKV = async <T>(key: string, pluginId: string): Promise<T | null> => {
+    console.log(`Getting KV entry for '${key}'`)
+    const client = await getClient();
+
+    const query = `SELECT pkey, pvalue from PluginKeyValueStore WHERE pkey = $1 AND pluginid = $2`;
+
+    const result = await client.query(query, [key, pluginId])
+        .finally(client.end);
+
+    if (!result.rows.length) {
+        return null;
+    }
+
+    const value = result.rows[0].pvalue?.toString();
+    try {
+        return JSON.parse(value);
+    } catch (e) {
+        return value;
+    }
+}
+
+export const setKV = async <T>(key: string, value: T, pluginId: string): Promise<void> => {
+    console.log(`Setting KV entry '${key}'`);
+
+    const existingEntry = await getKV(key, pluginId);
+
+    let query = 'INSERT INTO PluginKeyValueStore (pluginid, pkey, pvalue, expireat) VALUES ($1, $2, $3, $4)';
+    let values = [pluginId, key, value, 0];
+    if (existingEntry) {
+        query = 'UPDATE PluginKeyValueStore SET pvalue = $1 WHERE pkey = $2 AND pluginid = $3';
+        values = [value, key, pluginId];
+    }
+
+    const client = getClient();
+    await client.query(query, values)
+        .finally(client.end);
+}
+
+export const clearKVStoreForPlugin = async (pluginId: string) => {
```

Comment:
is there any way to do this server-side in order to avoid handling database directly?

Reply:
@trilopin I think this is the least obtrusive way to do this, because nothing needs to be changed to accomplish this. I see two alternatives:

- Support this through the plugin framework itself. We would need some sort of development endpoint or something to allow for this operation.
- Support this on the GitHub plugin itself. I'm not a big fan of adding something like this to a specific plugin project for development/testing services, but it's something we can do for sure.

I've also made a plugin to help with kv store management https://github.com/mickmister/mattermost-plugin-kv-backup-restore. The only way for it to "reach" into other plugin KV stores is to use SQL. It's possible to "masquerade" as a plugin to claim its ID and access the KV store that way, but that has proven to very inconvenient for the developer.

There are database operations being done for the webapp's Cypress tests, which is where I got the idea to do this https://github.com/mattermost/mattermost-server/blob/4a3c8536367f79f75ff026a9408d4aab84788cff/e2e-tests/cypress/tests/support/db_commands.ts#L6-L20